File: log_parser.py
import re
from dataclasses import dataclass

@dataclass
class LogEntry:
    timestamp: str
    level: str
    file: str
    function: str
    message: str
    original: str

    def to_file_format(self) -> str:
        """
        Returns the log string formatted for file saving.
        If it was a valid log, it returns the original line.
        If it was undefined, it constructs a formatted string with placeholders.
        """
        if self.timestamp == "UNDEFINED":
            # Construct a formatted string that mimics the standard ESP log format
            # [TIME][LEVEL][FILE:LINE] FUNCTION(): MESSAGE
            file_val = self.file if self.file != "UNDEFINED" else "UNDEFINED"
            return f"[UNDEFINED][{self.level}][{file_val}:0] UNDEFINED(): {self.message}"
        return self.original

class LogParser:
    # Regex based on: [TIME][LEVEL][FILE:LINE] FUNCTION(): MESSAGE
    # [   199] -> \[\s*(\d+)\]
    # [E] -> \[(.)\]
    # [Preferences.cpp:483] -> \[(.*?):(\d+)\]
    # getString(): -> \s*(.*?)\(\):
    # nvs_get_str len fail... -> \s*(.*)

    REGEX = re.compile(r'^\[\s*(\d+)\]\[([A-Z])\]\[(.*?):(\d+)\]\s*(.*?)\(\):\s*(.*)$')

    # Secondary Regex for System Logs: D Component: Message
    # ^([VDIWE]) -> Level
    # \s+
    # (.*?): -> Component (mapped to File)
    # \s+
    # (.*)$ -> Message
    REGEX_SYSTEM = re.compile(r'^([VDIWE])\s+(.*?):\s+(.*)$')

    @staticmethod
    def parse(line: str) -> LogEntry:
        line = line.strip()
        if not line:
            return None

        # Try Standard ESP Log Format
        match = LogParser.REGEX.match(line)
        if match:
            return LogEntry(
                timestamp=match.group(1),
                level=match.group(2),
                file=match.group(3),
                # group 4 is line number, we skip it as per requirements
                function=match.group(5),
                message=match.group(6),
                original=line
            )

        # Try System Log Format
        match_sys = LogParser.REGEX_SYSTEM.match(line)
        if match_sys:
             return LogEntry(
                timestamp="UNDEFINED",
                level=match_sys.group(1),
                file=match_sys.group(2),
                function="UNDEFINED",
                message=match_sys.group(3),
                original=line
            )

        # Fallback
        return LogEntry(
            timestamp="UNDEFINED",
            level="U",
            file="UNDEFINED",
            function="UNDEFINED",
            message=line,
            original=line
        )

File: lib/services/log_parser.dart
import 'package:serial_lens/models/log_entry.dart';

class LogParser {
  // Regex based on: [TIME][LEVEL][FILE:LINE] FUNCTION(): MESSAGE
  // Dart regex is almost identical to Python
  static final RegExp _regexStandard = RegExp(r'^\[\s*(\d+)\]\[([A-Z])\]\[(.*?):(\d+)\]\s*(.*?)\(\):\s*(.*)$');

  // Secondary Regex for System Logs: D Component: Message
  static final RegExp _regexSystem = RegExp(r'^([VDIWE])\s+(?:\(\d+\)\s+)?([^\[\]\s:]+):\s+(.*)$');

  static LogEntry parse(String line) {
    final trimmed = line.trim();
    if (trimmed.isEmpty) {
      // In Dart we typically return nullable, but to match existing flow
      // where we usually ignore empty lines before calling parse or handle nulls,
      // let's return a special entry or assume caller checks.
      // But adhering to the logic:
      return _createUndefined(trimmed);
    }

    final now = DateTime.now().millisecondsSinceEpoch / 1000.0;

    // Try Standard ESP Log Format
    final match = _regexStandard.firstMatch(trimmed);
    if (match != null) {
      return LogEntry(
        timestamp: match.group(1) ?? "UNDEFINED",
        level: match.group(2) ?? "U",
        file: match.group(3) ?? "UNDEFINED",
        // group 4 is line number, skipped
        function: match.group(5) ?? "UNDEFINED",
        message: match.group(6) ?? "",
        original: trimmed,
        arrivalTime: now,
      );
    }

    // Try System Log Format
    final matchSys = _regexSystem.firstMatch(trimmed);
    if (matchSys != null) {
      return LogEntry(
        timestamp: "UNDEFINED",
        level: matchSys.group(1) ?? "U",
        file: matchSys.group(2) ?? "UNDEFINED",
        function: "UNDEFINED",
        message: matchSys.group(3) ?? "",
        original: trimmed,
        arrivalTime: now,
      );
    }

    // Fallback
    return _createUndefined(trimmed, now);
  }

  static LogEntry _createUndefined(String line, [double? time]) {
    return LogEntry(
      timestamp: "UNDEFINED",
      level: "U",
      file: "UNDEFINED",
      function: "UNDEFINED",
      message: line,
      original: line,
      arrivalTime: time ?? (DateTime.now().millisecondsSinceEpoch / 1000.0),
    );
  }
}
